crates/ibc/src/core/context.rs

```diff
@@ -114,34 +93,46 @@ impl std::error::Error for RouterError {
 ///
 /// Trait used for the top-level [`validate`](crate::core::validate)
 pub trait ValidationContext: Router {
+    type ClientValidationContext;
+    type E: ClientExecutionContext;
```

Comment:
> Wouldn’t this lead to the unnecessary exposure of methods from one client (say client A) to another (client B)?

No; client A can only see the methods coming from its own trait, because that is the only trait bound that it will use in its implementation. For example, from the ADR,

```rust
impl<ClientValidationContext> ClientStateValidation<ClientValidationContext> for MyClientState
where
    ClientValidationContext: MyClientValidationContext,
{
    fn verify_client_message(
        &self,
        ctx: &ClientValidationContext,
        // ...
    ) -> Result<(), ClientError> { 
        // `get_resource_Y()` accessible through `ctx`
    }

    // ...
}
```
In this client's impl block, it can only use methods defined in `MyClientValidationContext`.

> Also, e.g. wherever there is an argument of type self::AnyClientState for a method of Client A, it potentially can accepts any other client variants defined under the AnyClientState.

By default, clients cannot do anything with `AnyClientState` because it is completely opaque. We solve that in the tendermint light client by adding [this trait bound](https://github.com/cosmos/ibc-rs/blob/746035633ff79003803cb0370553c8b9f0e4c4ed/crates/ibc/src/clients/ics07_tendermint/client_state.rs#L492), which allows us to convert `AnyClientState` into a tendermint `ClientState`. It doesn't matter that the enum has other variants (or that it's an enum at all); from the perspective of the light client, the only thing it can do with it is convert it into its own `ClientState`.

Reply:
For ppl implementing light clients it works so, though for the "light client developers", it is up to them whether to impose these restrictions. As in both the “impl block” and the `From<TmClientState>` bound you mentioned, they are done in the client-specific level (ics07) and not required by the bounds we have for the `ClientState` in ics02. 
However, I'm uncertain if there is a more efficient way to abstract and ensure that impls of each light client remain completely independent from one another.

Sounds like, there is an inevitable trade off between
“enabling light clients to introduce their own context methods” and “light client development complexity”